File: src/main/java/edu/cornell/cs/apl/viaduct/algebra/PartialOrder.java
package edu.cornell.cs.apl.viaduct.algebra;

/**
 * Like {@link Comparable}, but not all pairs of elements have to be ordered.
 *
 * @param <T> type of objects that this object may be compared to
 */
public interface PartialOrder<T> {
  /**
   * Returns true if {@code this} is ordered before {@code that}. It is not necessary that either
   * {@code this.lessThanOrEqualTo(that)} or {@code that.lessThanOrEqualTo(this)}.
   */
  boolean lessThanOrEqualTo(T that);
}

File: compiler/src/main/kotlin/io/github/aplcornell/viaduct/algebra/PartialOrder.kt
package io.github.apl_cornell.viaduct.algebra

/**
 * Like [Comparable], but not all pairs of elements have to be ordered.
 *
 * @param T the type of objects this object may be compared to.
 */
interface PartialOrder<T> {
    /**
     * Returns true if `this` is ordered before [that].
     *
     * It is not necessary that either `this.lessThanOrEqualTo(that)` or `that.lessThanOrEqualTo(this)`.
     */
    fun lessThanOrEqualTo(that: T): Boolean
}
